### PR TITLE
fix: constraints and proof logic

### LINKED ISSUE #19: Remove `expect`s on `submit_block_with_proofs`
We currently have some `expect` assertion that might cause the relay to crash.
- https://github.com/chainbound/helix/blob/bolt/crates/api/src/builder/api.rs#L741-L753
- https://github.com/chainbound/helix/blob/bolt/crates/api/src/builder/api.rs#L1571-L1597

### LINKED ISSUE #20: Review logs
We have too many noisy logs, especially in the context of delegations. They should be re-checked and removed or demoted to `trace` instead.
Also:
- we're missing logs when the inclusion proof is not available on `get_header_with_proofs`: https://github.com/chainbound/helix/blob/bolt/crates/api/src/proposer/api.rs#L505-L537

### PR DESCRIPTION
**Pull request by thedevbirb**
This PR is general fix/refactor over the constraints and the proof logic with better error handling. The main change is that now the `submit_blocks_with_proofs` endpoint fails if the bid contains no proofs when constraints are available.

Closes #19 
Closes #20 

## Release Notes

- **New Features**
	- Enhanced error handling for inclusion proofs and delegations, providing clearer feedback.
	- Improved logging detail across various methods for better tracking of operations.

- **Bug Fixes**
	- Streamlined handling of constraint submissions and payload validations, reducing potential errors.

- **Refactor**
	- Code organization and readability improvements across multiple files, ensuring consistent formatting and structure.

- **Documentation**
	- Updated error messages and logging to enhance clarity and usability for developers.

### CODE REVIEW

**Comment by thedevbirb** [comment]
Note: after this is merged we should upload a new image with tag `ghcr.io/chainbound/helix:v0.3.0-alpha.rc3`.

File: crates/api/src/builder/api.rs [hunk 1]
```diff
@@ -282,21 +275,10 @@ where
             .find(|duty| duty.slot == slot)
             .ok_or(BuilderApiError::ProposerDutyNotFound)?;
 
-        debug!(slot, duty = ?duty, "delegations -- found proposer duty");
-
         let pubkey = duty.entry.message.public_key.clone();
+        let delegations = Json(api.auctioneer.get_validator_delegations(pubkey).await?);
```
Comment (merklefruit) [inline crates/api/src/builder/api.rs:279]: why did you remove the logs here?

File: crates/api/src/builder/api.rs [hunk 2]
```diff
@@ -689,16 +671,42 @@ where
...
+                .as_slice()
+                .try_into()
+                .map_err(|e| {
+                    error!(error = %e, "failed to convert root to hash32");
+                    BuilderApiError::InternalError
+                })?;
+            let proofs = payload.proofs().ok_or(BuilderApiError::InclusionProofsNotFound)?;
+            let constraints_proofs: Vec<_> = constraints.iter().map(|c| &c.proof_data).collect();
+
+            verify_multiproofs(constraints_proofs.as_slice(), proofs, transactions_root).map_err(
```
Comment (merklefruit) [inline crates/api/src/builder/api.rs:690]: nice :)

File: crates/api/src/builder/api.rs [hunk 3]
```diff
@@ -689,16 +671,42 @@ where
...
+                payload.proposer_public_key(),
+                payload.block_hash(),
+                proofs,
+                &request_id,
+            )
+            .await?;
+
+            info!(request_id = %request_id, head_slot, "inclusion proofs verified and saved to auctioneer");
+        } else {
+            warn!(request_id = %request_id, "no constraints found for slot, proof verification is not needed");
```
Comment (merklefruit) [inline crates/api/src/builder/api.rs]: can we turn this level down to debug or info? it's warning of expected behaviour

File: crates/api/src/builder/api.rs [hunk 4]
```diff
@@ -689,16 +671,42 @@ where
...
+                    error!(error = %e, "failed to verify inclusion proofs");
+                    BuilderApiError::InclusionProofVerificationFailed(e)
+                },
+            )?;
 
-        // Verify inclusion proofs
-        if let Err(err) = api.verify_inclusion_proof(&payload, &constraints).await {
-            return Err(err);
+            // Save inclusion proof to auctioneer.
+            api.save_inclusion_proof(
```
Comment (merklefruit) [inline crates/api/src/builder/api.rs:698]: can you briefly explain why you moved `save_inclusion_proof` here as opposed to line 748?

**Review by merklefruit — APPROVED** [review]
Great!

File: crates/api/src/builder/api.rs (hunk 1 above)
Comment (thedevbirb) [reply crates/api/src/builder/api.rs:279]: There are already a ton of logs in Helix and I wanted to remove everything that's non-essential even at the debug level.
This log doesn't give anything helpful, if a proposer duty is found then we simply don't see the error which is enough for understanding if the code is working correctly.

Comment (merklefruit) [reply crates/api/src/builder/api.rs:279]: Gotcha, ty

File: crates/api/src/builder/api.rs (hunk 4 above)
Comment (thedevbirb) [reply crates/api/src/builder/api.rs:698]: Sure. In short we want to save proofs to cache only when:
- we have constraints
- we have proofs
- proofs are correct

Before we were already assuming proofs were available via `expect`, which may not be the case.

File: crates/api/src/builder/api.rs (hunk 3 above)
Comment (thedevbirb) [reply crates/api/src/builder/api.rs]: Right, I agree.

File: crates/api/src/builder/api.rs (hunk 4 above)
Comment (merklefruit) [reply crates/api/src/builder/api.rs:698]: Perfect

**Pull request merged**